File: matlab/tools/ecg/peak_det_simple.m
function [peaks, peak_indexes] = peak_det_simple(x, ff, flag, varargin)
%
% peak_det_simple - A basic R-peak detector
%
%   Syntax: [peaks, peak_indexes] = peak_det_simple(x, ff, flag, omit_close_peaks)
%
%   Inputs:
%       x: Vector of input data
%       ff: Approximate ECG beat-rate in Hertz
%       flag: Search for positive (flag=1), negative (flag=0) peaks, or
%           automatically using the polarity according to the dominant
%           peak/nadir median absolute amplitude
%       omit_close_peaks: omit close peaks after main peak detection
%           (true/1) or not(false/0). Default is 0
%
%   Outputs:
%       peaks: Vector of R-peak impulse train
%       peak_indexes: Vector of R-peak indexes
%
%   Revision History:
%       2006: First release
%       2023: Replaced deprecated version PeakDetection
%       2023: Added mode = 2 to automatically detect sign
%       2023: Added "&& sum(abs(x(index))) > 0" to avoid trivial all-zeros
%             peaks
%
%   Reza Sameni, Davood Fattahi, 2006-2023
%   The Open-Source Electrophysiological Toolbox
%   https://github.com/alphanumericslab/OSET

% Set default values for optional arguments
if nargin > 3 && ~isempty(varargin{1})
    omit_close_peaks = varargin{1};
else
    omit_close_peaks = 0;
end

N = length(x);

rng = floor(0.5 / ff);

peaks = zeros(1, N);
for j = 1 : N
    % Determine the index range for peak search
    if j > rng && j < N - rng
        index = j - rng : j + rng;
    elseif j > rng
        index = N - 2 * rng : N;
    else
        index = 1 : 2 * rng;
    end

    if max(x(index)) == x(j) && sum(abs(x(index))) > 0
        peaks(j) = 1;
    end
end

nadirs = zeros(1, N);
for j = 1 : N
    % Determine the index range for peak search
    if j > rng && j < N - rng
        index = j - rng : j + rng;
    elseif j > rng
        index = N - 2 * rng : N;
    else
        index = 1 : 2 * rng;
    end

    if min(x(index)) == x(j) && sum(abs(x(index))) > 0
        nadirs(j) = 1;
    end
end

switch flag
    case 1
        % No action needed, peaks has already been assigned
    case 0
        % Negative peaks has been explicitly requested
        peaks = nadirs; 
    case 2
        % Automatic mode, select the polarity according to the dominant
        % peak/nadir median absolute amplitude
        if abs(median(x(logical(peaks)))) < abs(median(x(logical(nadirs))))
            peaks = nadirs;
        end
end

% Remove fake peaks
if omit_close_peaks
    I = find(peaks);
    d = diff(I);
    peaks(I(d < rng)) = 0;
end
peak_indexes = find(peaks);

File: python/src/oset/ecg/peak_detection/peak_det_simple.py
import argparse

import numpy as np


def peak_det_simple(x, ff, flag, omit_close_peaks=0):
    """
    peak_det_simple - Internal R-peak detector function
    Syntax: peaks, peak_indexes = peak_det_simple(x, ff, flag, omit_close_peaks)

    Args:
        x (numpy.ndarray): vector of input data (ECG signal).
        ff (float): approximate ECG beat-rate in Hertz (normalized by the sampling frequency).
        flag (int): Search for positive (flag=1) or negative (flag=0) peaks.
        omit_close_peaks (int, optional): If non-zero (True), omit close peaks after main peak detection.
            Default is 0 (False).

      Returns:
          peaks: Vector of R-peak impulse train
          peak_indexes: Vector of R-peak indexes

    Revision History:
        July 2023: Translated to Python from Matlab (peak_det_simple.m)
        June 2024: Reworked the code to match the updated Matlab version

    Amulya Jain, 2023
    The Open-Source Electrophysiological Toolbox
    https://github.com/alphanumericslab/OSET
    """

    n = len(x)

    rng = int(np.floor(0.5 / ff))

    peaks = np.zeros(n)

    for j in range(1, n + 1):
        index = None
        # Determine the index range for peak search
        if rng < j < n - rng:
            index = slice(j - rng - 1, j + rng)
        elif j > rng:
            index = slice(j - rng - 1, n)
        else:
            index = slice(0, 2 * rng)
        if max(x[index]) == x[j - 1] and np.abs(x[index]).sum() > 0:
            peaks[j - 1] = 1

    nadirs = np.zeros(n)

    for j in range(1, n + 1):
        index = None
        # Determine the index range for peak search
        if rng < j < n - rng:
            index = slice(j - rng - 1, j + rng)
        elif j > rng:
            index = slice(n - 2 * rng - 1, n)
        else:
            index = slice(0, 2 * rng)
        if min(x[index]) == x[j - 1] and np.abs(x[index]).sum() > 0:
            nadirs[j - 1] = 1

    if flag == 0:
        peaks = nadirs
    elif flag == 2:
        if np.abs(np.median(x[peaks == 1])) < np.abs(np.median(x[nadirs == 1])):
            peaks = nadirs

    if omit_close_peaks:
        I = np.where(peaks)[0]
        d = np.diff(I)
        peaks[(I[d < rng])] = 0
    peak_indexes = np.where(peaks)[0] + 1
    return peaks, peak_indexes


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="""
    peak_det_simple - Internal R-peak detector function
    Syntax: peaks, peak_indexes = peak_det_simple(x, ff, flag, omit_close_peaks)

    Args:
        x (numpy.ndarray): vector of input data (ECG signal).
        ff (float): approximate ECG beat-rate in Hertz (normalized by the sampling frequency).
        flag (int): Search for positive (flag=1) or negative (flag=0) peaks.
        omit_close_peaks (int, optional): If non-zero (True), omit close peaks after main peak detection.
            Default is 0 (False).

      Returns:
          peaks: Vector of R-peak impulse train
          peak_indexes: Vector of R-peak indexes

    Revision History:
        July 2023: Translated to Python from Matlab (peak_det_simple.m)
        June 2024: Reworked the code to match the updated Matlab version

    Amulya Jain, 2023
    The Open-Source Electrophysiological Toolbox
    https://github.com/alphanumericslab/OSET
    """,
        formatter_class=argparse.RawTextHelpFormatter,
    )
    args = parser.parse_args()
